src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java

```diff
@@ -85,6 +87,67 @@ of this software and associated documentation files (the "Software"), to deal
 
 public class BuildData implements Serializable {
 
+    private static final BuildData EMPTY;
+
+    static {
+      try {
+          // exceptions below should never happen
+          // since constructor exits immediately if run is null
+          EMPTY = new BuildData(null, null);
+      } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(e);
+      } catch (IOException e) {
+          throw new RuntimeException(e);
+      }
+    }
+
+    private static final ThreadLocal<Set<Run<?, ?>>> BUILD_DATA_BEING_CREATED = ThreadLocal.withInitial(BuildData::identityHashSet);
+
+    private static <T> Set<T> identityHashSet() {
+        return Collections.newSetFromMap(new IdentityHashMap<>());
+    }
+
+    /**
+     * This is a workaround for a BuildData initialization issue.
+     * <p>
+     * Some of the fields in this class are initialized with the values obtained from a Run instance.
+     * Querying Run fields can trigger run initialization in some cases (e.g. loading run data from disk when resuming a run after Jenkins restart).
+     * During run initialization some logs are written.
+     * Writing logs, in turn, requires creating a TaskListener associated with the run.
+     * Creating the listener triggers initialization of DatadogTaskListenerDecorator, as the listener's logger needs to be decorated.
+     * The decorator creates another BuildData instance, whose creation starts the whole cycle from the beginning.
+     * As the result, the code enters an endless cycle of initializing BuildData while initializing BuildData, which terminates with a stack overflow.
+     * <p>
+     * The code below checks that BuildData for the provided run is already being created in the current thread.
+     * If that is the case, empty BuildData instance is returned for the nested calls in order to break the cycle.
+     * As a side effect, whatever logs are written while Run instance is being initialized will not be tagged with that run's data.
+     */
+    // TODO split BuildData fields that are used by DatadogWriter into a separate class whose initialization will not trigger run data load
+    @Nonnull
+    public static BuildData create(@Nullable Run<?, ?> run, @Nullable TaskListener listener) {
+        if (!BUILD_DATA_BEING_CREATED.get().add(run)) {
+            String runName = run != null ? run.getDisplayName() : null;
+            DatadogUtilities.severe(LOGGER, null, "BuildData creation is in progress for run " + runName + "; using empty data");
+            return BuildData.EMPTY;
```

Comment:
Although we return `EMPTY` here, we know that there's a call in the stack that will return the populated data? Is my assumption correct?

Reply:
Yes, exactly.
In short, the problem is that constructing a `BuildData` instance may trigger some run state to be initialised, and this results in Jenkins printing some logs to the console (something like "reading state for run #N from disk").
The plugin wants to send these logs to Datadog, just like any other logs. And it wants to tag each of these logs with the data of the associated run. But that requires a `BuildData`, so we enter an endless cycle.
By returning `EMPTY` we break this cycle. The downside is that logs triggered by `BuildData` init will miss the run info (they will only be tagged with the common Jenkins instance info).